[exporter/signalfx] Fix correlation timeout bug

crobert-1 opened the pull request:
**Description:** <Describe what has changed.>
The SignalFx Exporter supports correlations between metrics and traces. This means that when a trace occurs from a given host, with dimensions like service and environment, these dimensions can be used to query metrics for these traces. These correlations are supposed to timeout after the stale_service_timeout amount of time passes, but they weren't. This fix makes sure correlations timeout so that metrics aren't linked with stale trace data.

**Testing:**
Manual testing is working as expected, but it's currently impossible (from my understanding) to write tests for this functionality. The problem is that the tests would need to create both metrics and traces to send to the signalfx backend, but at that high of a level the correlation values couldn't be properly checked, as they're not available outside the local package.

crobert-1 committed:
Fix SignalFx Exporter correlation timeout bug

The SignalFx Exporter supports correlations between metrics
and traces. This means that when a trace occurs from a given
host, with dimensions like service and environment, these
dimensions can be used to query metrics for these traces.
These correlations are supposed to timeout after the
stale_service_timeout amount of time passes, but they
weren't. This fix makes sure correlations timeout so that
metrics aren't linked with stale trace data.

crobert-1 committed:
Update CHANGELOG

crobert-1 committed:
Add doc file for timeutils package

pmcollins commented on internal/coreinternal/timeutils/ticker_helper.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+// Ensure PolicyTicker implements TTicker interface
+var _ TTicker = (*PolicyTicker)(nil)
```
Can we move this assertion up to right above or below the PolicyTicker struct definition?

crobert-1 replied:
I've moved it right under the struct definition.

pmcollins reviewed (approved):
LGTM, just a couple of nits. Thanks for fixing this!

crobert-1 committed:
Changes requested in PR

- Add public type comment
- Move implementing checker right under struct declaration
- Add spaces between functions
- Fix failing PR checks: Delay in tests failed in automation,
add license block in test file.

crobert-1 commented on internal/coreinternal/timeutils/ticker_helper.go:
```diff
@@ -0,0 +1,58 @@
+		}
+	}()
+}
+func (pt *PolicyTicker) OnTick() {
```
Added, thanks!

crobert-1 committed:
Address racing test bug

crobert-1 committed:
Fix tests that had a race in them

GitHub automation was failing unit tests because of a race
detected. The solution was to add a mutex to checking and
setting the variable in question.

crobert-1 committed:
Address lint and unit test failure

crobert-1 committed:
Fix linting failure and test timing

crobert-1 commented:
_Failing tests are unrelated to my changes_

dmitryax reviewed (approved):
LGTM

dmitryax committed:
Merge branch 'main' into sfx_export_correlate_fix
